### ISSUE TITLE
ReadyToRun executable crashes at Marshal.StructureToPtr when trying to Marshal an Inherited class

### ISSUE THREAD

**Issue by dhiru1602** [dotnet/runtime#63106]
### Description

The following sample code crashes when ReadyToRun is enabled in the Publish Settings.

```
using System.Runtime.InteropServices;

byte[] testString = (new MSG1()
{
    e = 20
}).GetBuffer();

Console.WriteLine(Convert.ToHexString(testString));
Console.ReadLine();

[StructLayout(LayoutKind.Sequential, Pack = 1)]
class IPacket
{    internal byte[] GetBuffer()
    {
        byte[] arr = new byte[GetSize()];
        IntPtr hglobal = Marshal.AllocHGlobal(GetSize());
        Marshal.StructureToPtr(this, hglobal, true);
        Marshal.Copy(hglobal, arr, 0, arr.Length);
        Marshal.FreeHGlobal(hglobal);

        return arr;
    }

    internal ushort GetSize()
    {
        return (ushort)Marshal.SizeOf(this);
    }
}

[StructLayout(LayoutKind.Sequential, Pack = 1)]
internal class MSG_HEADER : IPacket
{
    internal MSG_HEADER() { }
    internal MSG_HEADER(uint _b, byte _c, byte _d)
    {
        b = _b;
        c = _c;
        d = _d;
        a = GetSize();
    }
    internal UInt32 a;
    internal UInt32 b;
    internal Byte c;
    internal Byte d;
}

[StructLayout(LayoutKind.Sequential, Pack = 1)]
internal class MSG1 : IPacket
{
    internal MSG1()
    {
        MsgHeader = new MSG_HEADER
        {
            a = GetSize(),
            b = 0x1800,
            c = 0x01,
            d = 0xE0
        };
    }
    internal MSG_HEADER MsgHeader;
    internal byte e;
}
```

### Reproduction Steps

Use the above code and publish the executable with **Enable ReadyToRun compilation** enabled. The executable crashes with an Object Reference not set error.

### Expected behavior

The following is printed when ReadyToRun compilation is disabled and the program executes as intended.

`0B0000000018000001E014`

### Actual behavior

The executable crashes with the following error when ReadyToRun compilation is enabled in Publish Settings

```
Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Runtime.InteropServices.Marshal.StructureToPtr(Object structure, IntPtr ptr, Boolean fDeleteOld)
   at System.Runtime.InteropServices.Marshal.StructureToPtr[T](T structure, IntPtr ptr, Boolean fDeleteOld)
```

### Regression?

Executable works fine on .NET Core Runtine 5.0.7 with ReadyToRun compilation enabled
Issue occurs on .NET Core 6.0 with ReadyToRun compilation enabled.

### Known Workarounds

Disabling ReadyToRun compilation fixes the executable

### Configuration

.NET Version: 6.0.101
Windows Server 2019
x64

Publish Settings:
Configuration: Release
Target Framework: net6.0
Target Runtime: win-x64
Produce Single File
Enable ReadyToRun compilation

**Comment by AaronRobinsonMSFT** [comment]
@jkoritzinsky Didn't we hit something similar recently?

**Comment by mangod9** [comment]
@AaronRobinsonMSFT @jkoritzinsky is this something we need to look into for 7?

**Comment by mangod9** [comment]
Assigning to Jeremy for now and leaving it in 7.0. Please move based on risk. 

@trylek fyi.

**Comment by jkoritzinsky** [comment]
I finally root-caused this bug:

This is a class-layout bug. Crossgen2 thinks that the layout should be one thing, but CoreCLR treats it differently. This only shows up in the `Marshal.StructureToPtr` call since that method calls into CoreCLR itself, which generates the structure marshalling code based on CoreCLR's idea of the type layout. I'll put up a PR to fix it, but due to it touching type layout, I think this is better suited to be fixed early in .NET 8 than late in .NET 7 (as this type of fix generally has higher risk)

**Comment by mangod9** [comment]
Thanks for investigating. Agree that we should move to 8.

**Comment by jkotas** [comment]
The type layout bugs are hard to debug, and fixing them is typically R2R breaking change. We took R2R breaking change for 7.0. We should get all type load bugs fixed for 7.0, so that we do not have to take R2R breaking change in 8.0.
